src/graph.rs

```diff
@@ -1885,7 +1962,129 @@ impl<'a, 'graph> Builder<'a, 'graph> {
+        for (specifier, _, maybe_range) in infos {
+          if !self.graph.module_slots.contains_key(&specifier) {
+            self.graph.module_slots.insert(
+              specifier.clone(),
+              ModuleSlot::Err(ModuleGraphError::LoadingErr(
+                specifier,
+                maybe_range,
+                Arc::new(anyhow::anyhow!("{}", err)),
+              )),
+            );
+          }
+        }
+      } else {
+        for (specifier, npm_ref, maybe_range) in infos {
+          if !self.graph.module_slots.contains_key(&specifier) {
+            if let Some(npm_resolver) = &self.npm_resolver {
+              // todo: cache resolutions to ensure they always resolve the same
+              let resolution = npm_resolver.resolve_npm(&npm_ref.req);
+              match resolution {
+                Ok(pkg_id) => {
+                  specifier_resolutions
+                    .insert(specifier.clone(), pkg_id.clone());
+                  let pkg_id_ref = NpmPackageNvReference {
+                    nv: pkg_id,
+                    sub_path: npm_ref.sub_path,
+                  };
+                  let resolved_specifier = pkg_id_ref.as_specifier();
+                  if resolved_specifier != specifier {
+                    self
+                      .graph
+                      .redirects
+                      .insert(specifier, resolved_specifier.clone());
+                  }
+                  self.graph.module_slots.insert(
+                    resolved_specifier.clone(),
+                    ModuleSlot::Module(Module::Npm(NpmModule {
+                      specifier: resolved_specifier,
+                      nv_reference: pkg_id_ref,
+                    })),
+                  );
+                }
+                Err(err) => {
+                  self.graph.module_slots.insert(
+                    specifier.clone(),
+                    ModuleSlot::Err(ModuleGraphError::ResolutionError(
+                      ResolutionError::ResolverError {
+                        error: Arc::new(err),
+                        specifier: specifier.to_string(),
+                        // this should always be set,
+                        range: maybe_range.unwrap_or_else(|| Range {
+                          specifier,
+                          start: Position::zeroed(),
+                          end: Position::zeroed(),
+                        }),
+                      },
+                    )),
+                  );
+                }
+              }
+            } else {
+              self.graph.module_slots.insert(
+                specifier.clone(),
+                ModuleSlot::Err(ModuleGraphError::LoadingErr(
+                  specifier,
+                  maybe_range,
+                  Arc::new(anyhow::anyhow!(
+                    "npm specifiers are not supported in this environment"
+                  )),
+                )),
+              );
+            }
+          }
+        }
+      }
```

Comment:
This is very nice, consider factoring the whole body of `while` into a helper function, could be much easier to grep with fewer indents

Reply:
I was thinking about that, but got lazy. Will update soon

Reply:
It's a lot of work to do this properly. I'm going to skip on this for now and add a todo